app/views/problems/submit.html.erb

```diff
@@ -8,33 +8,48 @@
     </ul>
   </div>
 <% end %>
+<h2>Upload file</h2>
 <%= form_for @submission, :html => {:multipart => true}, :url => submit_problem_path(@problem) do |f| %>
-  <div class="field">
+  <div class="subfield">
     Language: 
     <%= f.select :language_id, Language.submission_options %>
   </div>
   <div class="field">
     Source file: 
     <%= f.file_field :source_file %>
   </div>
-  <div class="actions">
+  <div class="subactions">
     <%= f.submit 'Submit' %>
   </div>
 <% end %>
 <% if policy(@problem).submit_source? %>
-  <center><b>or submit source directly:</b></center>
+  <br>
+  <hr>
+  <br>
+  <h2>Or paste code</h2>
   <%= form_for @submission, :url => submit_problem_path(@problem) do |f| %>
-    <div class="field">
+    <div class="subfield">
       Language: 
       <%= f.select :language_id, Language.submission_options %>
     </div>
+    <div class="subactions">
+      <%= f.submit 'Submit' %>
+    </div>
     <div class="field">
       <%= f.label :source %><br />
       <%= f.text_area :source, :style => "width: 100%", :cols => 60 %>
     </div>
-    <div class="actions">
-      <%= f.submit 'Submit' %>
-    </div>
   <% end %>
 <% end %>
-
+<br>
+<hr>
+<br>
+<h2> Language details </h2>
+<p>
+<ul>
+  <li>C++ is compiled with g++ from GCC 9.2.1</li>
+  <li>C is compiled with gcc from GCC 9.2.1</li>
+  <li>Java uses javac 1.8 and OpenJDK 9</li>
+  <li>Haskell is compiled with GHC 7.10.3</li>
```

Comment:
I don't think we should claim such precise versions when they can change in a routine 'apt upgrade'.
How about we only specify the major versions? GCC 9 and GHC 7.10 should be detailed enough for most people.
Another idea is to modify the compilation commands so they also print the compiler's version...

Reply:
Agreed. I like the compilation command idea too. I'll leave this PR open for now to see if anyone wants to give implementing that a go

Reply:
Maybe split off the readability improvements into a separate PR so they don't get held up by the language details?
I think they are pretty important, since the textarea input ("paste") is now visible to all users and in my experience the previous UI can be confusing.

Reply:
Done